textx/scoping/providers.py

```diff
@@ -572,18 +645,23 @@ def get_reference_propositions(self, obj, attr, name_part):
             raise TextXError(
                 "expected path to list in the model ({})".format(
                     self.path_to_container_object))
-        obj_list = filter(
-            lambda x: textx_isinstance(x, attr.cls) and
-            x.name.find(name_part) >= 0, obj_list)
 
-        return list(obj_list)
+        name_obj_tuples = _objects_list_to_tuples_of_name_and_objects(
+            obj_list, self.name_resolver_logic, name_part, attr.cls
+        )
+
+        return name_obj_tuples
 
     def __call__(self, obj, attr, obj_ref):
-        lst = self.get_reference_propositions(obj, attr, obj_ref.obj_name)
-        if type(lst) is Postponed:
-            return lst
-        if len(lst) > 0:
-            return lst[0]
+        lst_raw = self.get_reference_propositions(
+            obj, attr, obj_ref.obj_name)
+        if type(lst_raw) is Postponed:
+            return Postponed()
+        lst = _filter_postponed_from_tuples_of_name_and_objects(lst_raw)
+        if len(lst_raw) > len(lst):
```

Comment:
I see here that the list is necessary

Reply:
Note: Once we drop python 2.7, we may be able to changesome of of the "get_list_of" to (recursive) generators. I think there are restrictions in 2.7 for that (recursive generators)...

Reply:
This code could be optimized. It is implemented that way to be consistent (+somehow readable) with the spec (in scoping.md).